rfc/rfc-56/rfc-56.md

```diff
@@ -0,0 +1,231 @@
+1. DirectWriteMarkers: individual marker file corresponding to each data file is directly created by the writer.
+2. TimelineServerBasedWriteMarkers: marker operations are all handled at the timeline service which serves as a proxy
+
+Therefore, for different types of Marker we have, we must implement the corresponding check marker conflict logic. 
+Here we expand the existing `ConflictResolutionStrategy` interface to ensure the scalability of checking marker conflict.
+
+![](flow1.png)
+
+In this design, we provide `DirectMarkerWithTransactionConflictResolutionStrategy` and 
+`SimpleDirectMarkerConflictResolutionStrategy` for DirectWriteMarkers to perform corresponding conflict detection and 
+conflict resolution. And we provide `AsyncTimelineMarkerConflictResolutionStrategy` for TimelineServerBasedWriteMarkers 
+to perform corresponding conflict detection and conflict resolution
+
+#### DirectWriteMarkers related strategy
+
+##### DirectMarkerWithTransactionConflictResolutionStrategy
+
+![](figure2.png)
+
+As for current strategy showed by figure 2, we need to pay attention to the following two details:
+
+
+Firstly, we will use the existing transaction mechanism to lock at the <strong>PartitionPath + "/" + fileID</strong>
+level before checking conflicts and before creating marker files. 
+Take ZK locker as an example => `LockConfiguration.ZK_LOCK_KEY_PROP_KEY, partitionPath + "/" + fileId`. 
+That is to say, for the same partitionPath/fileId, only one writer can check and create the corresponding marker file 
+at the same time.
+
+
+In addition, during checking conflicts, we do not need to list all marker files in all directories, we can prune based 
+on the current partition Path and fileID to avoid unnecessary list operations.
+
+##### SimpleDirectMarkerConflictResolutionStrategy
+
+Compared with `DirectMarkerWithTransactionConflictResolutionStrategy`, the current strategy drops the steps of 
+Transaction including new Transaction, begin transaction and end transaction. The advantages are that the checking 
+speed is faster, and multiple writers will not affect each other. But the downside is that the conflict may be 
+underreported. 
+
+
+For example, when two writers detect PartitionPath1 + fileID1 at the same time, 
+both writers will pass conflict checking and successfully create a marker, which is an underreported. 
+If there is an underreported, the conflict can only be found in the pre-commit conflict checking and fail writing. 
+This will lead to waste of resources, but will not compromise the correctness of the data.
+
+#### TimelineServerBasedWriteMarkers related strategy
+
+##### AsyncTimelineMarkerConflictResolutionStrategy
+
+This design will register a new Marker API called `/v1/hoodie/marker/check-marker-conflict` on timeline server.
+
+![](figure3.png)
+
+As shown on figure3. For the client side, it will call this check-marker-conflict api to get the latest value 
```

Comment:
Nice catch Ethan. I'd like more discuss here. 
A separate api is added here for better abstraction. Because the follow-up may not only support async checking, but also support loop polling based on each writer's time line server. And coupling checking and create together may lose some scalability.